File: lib/features/player/domain/entities/watch_history_entry.dart
import 'package:flutter/foundation.dart';

import '../../../../core/constants.dart';

/// A playback history entry for resume support.
///
/// Domain entity — pure Dart, no infrastructure
/// dependencies. Replaces Drift's DbWatchHistoryData.
@immutable
class WatchHistoryEntry {
  const WatchHistoryEntry({
    required this.id,
    required this.mediaType,
    required this.name,
    required this.streamUrl,
    this.posterUrl,
    this.seriesPosterUrl,
    this.positionMs = 0,
    this.durationMs = 0,
    required this.lastWatched,
    this.seriesId,
    this.seasonNumber,
    this.episodeNumber,
    this.deviceId,
    this.deviceName,
    this.profileId,
    this.sourceId,
  });

  /// Unique history entry identifier.
  final String id;

  /// Content type: "channel", "movie", or "episode".
  final String mediaType;

  /// Display name of the watched content.
  final String name;

  /// Stream URL that was playing.
  final String streamUrl;

  /// Poster image URL for the history list.
  final String? posterUrl;

  /// Series poster fallback URL.
  final String? seriesPosterUrl;

  /// Playback position in milliseconds.
  final int positionMs;

  /// Total duration in milliseconds.
  final int durationMs;

  /// When this content was last watched.
  final DateTime lastWatched;

  /// Parent series ID (for episodes).
  final String? seriesId;

  /// Season number (for episodes).
  final int? seasonNumber;

  /// Episode number (for episodes).
  final int? episodeNumber;

  /// Device ID for multi-device sync.
  final String? deviceId;

  /// Human-readable device name.
  final String? deviceName;

  /// Profile ID that watched this content.
  final String? profileId;

  /// Source ID that provided this content (for multi-source tracking).
  final String? sourceId;

  /// Playback progress as a fraction (0.0 – 1.0).
  double get progress => durationMs > 0 ? positionMs / durationMs : 0.0;

  /// Whether playback is nearly complete (>= 95%).
  bool get isNearlyComplete => progress >= kCompletionThreshold;

  /// Short episode label for series items, e.g. "S2 E5".
  ///
  /// Returns null when [seasonNumber] or [episodeNumber] is missing.
  String? get episodeLabel {
    if (seasonNumber == null || episodeNumber == null) return null;
    return 'S$seasonNumber E$episodeNumber';
  }

  /// Returns a display label like "S2 E5 · 15m left" or "23m left".
  String get remainingLabel {
    final parts = <String>[];
    final ep = episodeLabel;
    if (ep != null) parts.add(ep);
    if (durationMs > 0 && positionMs < durationMs) {
      final remainMs = durationMs - positionMs;
      final remainMin = (remainMs / 60000).ceil();
      parts.add('${remainMin}m left');
    }
    return parts.join(' · ');
  }

  WatchHistoryEntry copyWith({
    String? id,
    String? mediaType,
    String? name,
    String? streamUrl,
    String? posterUrl,
    String? seriesPosterUrl,
    int? positionMs,
    int? durationMs,
    DateTime? lastWatched,
    String? seriesId,
    int? seasonNumber,
    int? episodeNumber,
    String? deviceId,
    String? deviceName,
    String? profileId,
    String? sourceId,
  }) {
    return WatchHistoryEntry(
      id: id ?? this.id,
      mediaType: mediaType ?? this.mediaType,
      name: name ?? this.name,
      streamUrl: streamUrl ?? this.streamUrl,
      posterUrl: posterUrl ?? this.posterUrl,
      seriesPosterUrl: seriesPosterUrl ?? this.seriesPosterUrl,
      positionMs: positionMs ?? this.positionMs,
      durationMs: durationMs ?? this.durationMs,
      lastWatched: lastWatched ?? this.lastWatched,
      seriesId: seriesId ?? this.seriesId,
      seasonNumber: seasonNumber ?? this.seasonNumber,
      episodeNumber: episodeNumber ?? this.episodeNumber,
      deviceId: deviceId ?? this.deviceId,
      deviceName: deviceName ?? this.deviceName,
      profileId: profileId ?? this.profileId,
      sourceId: sourceId ?? this.sourceId,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is WatchHistoryEntry &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => Object.hash(runtimeType, id);

  @override
  String toString() => 'WatchHistoryEntry($name, $mediaType)';
}

File: src/Crispy.Application/Player/Models/WatchHistoryEntry.cs
namespace Crispy.Application.Player.Models;

/// <summary>
/// Discriminates the type of media in a watch history entry (PLR-48).
/// </summary>
public enum MediaType
{
    /// <summary>Live TV channel.</summary>
    Channel = 0,

    /// <summary>VOD movie.</summary>
    Movie = 1,

    /// <summary>Individual episode of a series.</summary>
    Episode = 2,
}

/// <summary>
/// Records a single playback session for any content type.
/// ID is the SHA-256 hash of the stream URL, first 8 bytes hex-encoded (PLR-47).
/// </summary>
public class WatchHistoryEntry
{
    /// <summary>SHA-256(streamUrl)[0..8] hex — stable across renames (PLR-47).</summary>
    public required string Id { get; set; }

    /// <summary>Content type discriminator (PLR-48).</summary>
    public MediaType MediaType { get; set; }

    /// <summary>Display name for the content.</summary>
    public required string Name { get; set; }

    /// <summary>The actual stream URL used for playback.</summary>
    public required string StreamUrl { get; set; }

    /// <summary>Poster/thumbnail URL.</summary>
    public string? PosterUrl { get; set; }

    /// <summary>Series-level poster (used when MediaType == Episode).</summary>
    public string? SeriesPosterUrl { get; set; }

    /// <summary>Playback position in milliseconds.</summary>
    public long PositionMs { get; set; }

    /// <summary>Total duration in milliseconds (0 for live channels).</summary>
    public long DurationMs { get; set; }

    /// <summary>UTC timestamp of the last watch activity.</summary>
    public DateTimeOffset LastWatched { get; set; }

    /// <summary>Parent series ID (null for non-episode content).</summary>
    public string? SeriesId { get; set; }

    /// <summary>Season number within the series (null for non-episode content).</summary>
    public int? SeasonNumber { get; set; }

    /// <summary>Episode number within the season (null for non-episode content).</summary>
    public int? EpisodeNumber { get; set; }

    /// <summary>Device identifier for multi-device sync.</summary>
    public required string DeviceId { get; set; }

    /// <summary>Human-readable device name.</summary>
    public required string DeviceName { get; set; }

    /// <summary>Profile this entry belongs to.</summary>
    public required string ProfileId { get; set; }

    /// <summary>Source (provider) this content came from.</summary>
    public required string SourceId { get; set; }

    // -------------------------------------------------------------------------
    // Computed properties (PLR-49)
    // -------------------------------------------------------------------------

    /// <summary>
    /// Playback progress as a fraction 0.0–1.0 (PLR-49).
    /// Returns 0.0 when DurationMs is 0 (live channels).
    /// </summary>
    public double Progress => DurationMs > 0 ? (double)PositionMs / DurationMs : 0.0;

    /// <summary>
    /// True when playback has started but is not yet complete (PLR-44).
    /// Progress must be > 0 and &lt; 0.95 (95 % threshold marks "watched").
    /// </summary>
    public bool IsInProgress => Progress > 0.0 && Progress < 0.95;
}
